### ISSUE TITLE
Building GO version

### ISSUE THREAD

**Issue by aborruso** [johnkerl/miller#373]
> That said, if you would be OK with cloning the Miller repo, `cd go`, and `./build`, it would be useful to me to see how that works for you.

When I run it in my debian I have

```
./build: line 12: check: command not found
```

I have `go version go1.12 linux/amd64`.

Thank you

**Comment by johnkerl** [comment]
Ahh OK. TL;DR I have `.` in my `$PATH` and I should not expect that.

I'll push a change.

**Comment by johnkerl** [comment]
OK please try `git pull` in your clone now. (https://github.com/johnkerl/miller/commit/7bb40e2c2b767038ea1fc18de01655c6d101de7f)

**Comment by aborruso** [comment]
I probably won't be able to do it until Monday. I'm sorry.

Thank you very much



Il ven 18 set 2020, 20:45 John Kerl <notifications@github.com> ha scritto:

**Comment by aborruso** [comment]
@johnkerl I just tried and after a while it appears to me what you see below.

I think it's not so good.

![image](https://user-images.githubusercontent.com/30607/93660464-8b05e080-fa4f-11ea-905a-3c47ab484a09.png)

**Comment by johnkerl** [comment]
OK here is an executable as of this moment. Caveats:

* As above, this is very much partial.
* `mlr -l` gives only `cat cut group-by group-like head label nothing put rename sort tac tail`
* At present you have to do `mlr put 'filter NR > 10'` since `mlr filter 'NR > 10'` doesn't exist yet
* `mlr --help` advertises some things the Go port doesn't do yet
* DKVP and NIDX will handle double-quoting but they don't yet

In short ... I don't know if this will be of much value to you just yet, but, I'm happy to hear whatever your thoughts are! :)
[mlr.go.linux_x86_64.gz](https://github.com/johnkerl/miller/files/5258841/mlr.go.linux_x86_64.gz)

**Comment by aborruso** [comment]
@johnkerl I have tested `echo "a=this is a test" | mlr_go put '$b=sub($a,"a","")'`, and I have

```
CST BuildFunctionCallsiteNode: function name not found: sub
```

If I run `hellp` I have `sub` listed in the funcions list. Am I doing something wrong?

**Comment by johnkerl** [comment]
There is no `sub` yet, and in general, nothing with regular expressions yet.

**Comment by johnkerl** [comment]
`mlrgo` doesn't print a functions list yet ... are you doing `mlr --help` ?

**Comment by aborruso** [comment]
> `mlrgo` doesn't print a functions list yet ... are you doing `mlr --help` ?

you're right, sorry, I misunderstood

**Comment by johnkerl** [comment]
@aborruso as of very recently I have a Windows laptop to experiment with :)

Build instructions for Windows will be improving :)

**Comment by aborruso** [comment]
@johnkerl My OS is Windows, but my data OS is Linux.

I use Miller only inside Linux. Then for me it would be great to build easily Go Miller inside my debian distro.

Having said that, I think it's extraordinary to have Miller for Win, because it's a rare gem

**Comment by johnkerl** [comment]
@aborruso OK cool

Well, for Linux, at this point the module structure is that you just do

```
cd miller/go
go build mlr.go
```

and that's all.

(I have a wrapper script around that called `build` which also invokes regression tests and such but the core of it is `go build mlr.go`.)

(The build story for Windows is equally simple -- in both cases, as long as you have already installed the Go tools. However there are CR/LF and single-quote issues in the Windows port that I simply haven't touched yet.)

_However_

**Comment by johnkerl** [comment]
@aborruso the aforementioned CR/LF and single-quote issues have been dealt with.

There's more to do of course but if you install Go and then `go build mlr.go` within the `miller\go` subdirectory you should get a `mlr.exe`.

**Comment by aborruso** [comment]
> cd miller/go
> go build mlr.go

Now I have it in my debian and it works. I will test it, I'm very happy and now I can use lazy regex :)

If I run

```
echo "a=this is @ a @ test" | mlr put '$b=sub($a,"^.+?@","")'
```

I have `a=this is @ a @ test,b= test`.

But if I run 

```
echo "a=this is @ a @ test" | mlrgo put '$b=sub($a,"^.+?@","")'
```

I have `a=this is @ a @ test,b= a @ test`.

You are great John

**Comment by johnkerl** [comment]
@aborruso I will close this I think -- between http://johnkerl.org/miller-releases/miller-head/docs6/_build/html/build.html and the option of visiting https://github.com/johnkerl/miller/actions, selecting the latest build for `main`, and downloading the binary for Linux / MacOS / Windows I think we are covered -- ? Please let me know if this should remain open.

**Comment by aborruso** [comment]
> visiting https://github.com/johnkerl/miller/actions, selecting the latest build for `main`, and downloading the binary for Linux / MacOS / Windows I think we are covered

I feel stupid :)

When I visit https://github.com/johnkerl/miller/actions, where are the last builds to download? Thank you

**Comment by johnkerl** [comment]
Sorry @aborruso that was unclear!

At https://github.com/johnkerl/miller/actions click on the latest successful build for `main`:

<img width="872" alt="Screen Shot 2021-06-19 at 10 44 53 AM" src="https://user-images.githubusercontent.com/2008794/122646180-e999e100-d0eb-11eb-8a7c-d8c6fbeb6f63.png">

Then scroll down to where the "Artifacts" are:

<img width="883" alt="Screen Shot 2021-06-19 at 10 45 18 AM" src="https://user-images.githubusercontent.com/2008794/122646191-f585a300-d0eb-11eb-8ad6-d1981bd7be4f.png">

**Comment by aborruso** [comment]
@johnkerl bombastic, thank you
